core/client/fs/src/main/java/alluxio/client/file/cache/filter/ConcurrentClockCuckooFilter.java

```diff
@@ -0,0 +1,1070 @@
+    int maxRetryNum = 1;
+    boolean done = false;
+    while (maxRetryNum-- > 0) {
+      if (cuckooInsert(b1, b2, fp, pos)) {
+        done = true;
+        pos.mStatus = CuckooStatus.OK;
+        break;
+      }
+    }
+    return done;
+  }
+
+  /**
+   * Try find an empty slot for the item. Assume already held the lock of buckets i1 and i2.
+   *
+   * @param b1 the first bucket
+   * @param b2 the second bucket
+   * @param fp the fingerprint
+   * @param pos the position
+   * @return true iff found an empty slot (stored in pos); false otherwise
+   */
+  private boolean cuckooInsert(int b1, int b2, int fp, TagPosition pos) {
+    TagPosition pos1 = new TagPosition();
+    TagPosition pos2 = new TagPosition();
+    // try find b1 and b2 firstly
+    if (!tryFindInsertBucket(b1, fp, pos1)) {
+      pos.setStatus(CuckooStatus.FAILURE_KEY_DUPLICATED);
+      return false;
+    }
+    if (!tryFindInsertBucket(b2, fp, pos2)) {
+      pos.setStatus(CuckooStatus.FAILURE_KEY_DUPLICATED);
+      return false;
+    }
+    if (pos1.getTagIndex() != -1) {
+      pos.setBucketAndSlot(b1, pos1.getTagIndex());
+      pos.setStatus(CuckooStatus.OK);
+      return true;
+    }
+    if (pos2.getTagIndex() != -1) {
+      pos.setBucketAndSlot(b2, pos2.getTagIndex());
+      pos.setStatus(CuckooStatus.OK);
+      return true;
+    }
+    // then BFS search from b1 and b2
+    boolean done = runCuckoo(b1, b2, fp, pos);
+    if (done) {
+      // avoid another duplicated key is inserted during runCuckoo.
+      if (mTable.findTagInBuckets(b1, b2, fp)) {
+        pos.setStatus(CuckooStatus.FAILURE_KEY_DUPLICATED);
+        return false;
+      } else {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  /**
+   * Try find an empty slot for the item. Assume already held the lock of buckets i1 and i2.
+   *
+   * @param b1 the first bucket
+   * @param b2 the second bucket
+   * @param fp the fingerprint
+   * @param pos the position
+   * @return true iff found an empty slot (stored in pos); false otherwise
+   */
+  private boolean runCuckoo(int b1, int b2, int fp, TagPosition pos) {
+    mLocks.unlockTwoWrite(b1, b2);
+    int maxPathLen = MAX_BFS_PATH_LEN;
+    CuckooRecord[] cuckooPath = new CuckooRecord[maxPathLen];
+    for (int i = 0; i < maxPathLen; i++) {
+      cuckooPath[i] = new CuckooRecord();
+    }
+    boolean done = false;
+    while (!done) {
+      int depth = cuckooPathSearch(b1, b2, fp, cuckooPath);
+      if (depth < 0) {
+        break;
+      }
+      if (cuckooPathMove(b1, b2, fp, cuckooPath, depth)) {
+        pos.setBucketAndSlot(cuckooPath[0].mBucket, cuckooPath[0].mSlot);
+        pos.setStatus(CuckooStatus.OK);
+        done = true;
+      }
+    }
+    if (!done) {
+      // NOTE: since we assume holding the locks of two buckets before calling this method,
+      // we keep this assumptions after return.
+      mLocks.lockTwoWrite(b1, b2);
```

Comment:
any chance to do the unlock and lock on the caller side?   because the caller might be aware of the lock status